typing-break/main.c

```diff
@@ -98,26 +77,6 @@ main (int argc, char *argv[])
 		return 0;
 	}
 
-#ifndef HAVE_APP_INDICATOR
-	if (!no_check && !have_tray ()) {
```

Comment:
should we have something similar for appindicator?  It should be doable checking whether there's a status notifier host registered.  Not saying it's needed, just a thought.

Reply:
done

Reply:
Do you think checking for a tray makes sense?  For us it does because the notification area does both SN host and system tray (on x11, on Wyaland there won't be a system tray I guess), but a better check would be whether there's the right DBus thingy at the right place -- not that I can tell you how to do that, but some inspiration from the SN host in notification-area might possibly help.

Then again, I'm not entirely sure how important that warning is in real life.

Reply:
I plan to use Dbus to check if the notification area exists, but I have encountered a problem，I'm not sure if it's my usage issue
```
dbus-send --session --print-reply=literal --dest=org.kde.StatusNotifierWatcher  /StatusNotifierWatcher org.freedesktop.DBus.Introspectable.Introspect
```
https://gitlab.gnome.org/GNOME/glib/-/issues/3150

Reply:
I barely know anything about DBus, let alone GDbus, but this seems to work (indeed `g_dbus_proxy_new_for_bus_sync()` don't seem to want to fail, but it's not necessarily a problem here):

```c
#include <stdio.h>
#include <glib.h>
#include <gio/gio.h>

static gboolean
is_status_notifier_host_available (void)
{
    const gchar *service = "org.kde.StatusNotifierWatcher";
    const gchar *path = "/StatusNotifierWatcher";
    const gchar *interface = "org.kde.StatusNotifierWatcher";
    const gchar *property = "IsStatusNotifierHostRegistered";

    g_autoptr (GDBusProxy) proxy = NULL;
    g_autoptr (GError) error = NULL;

    proxy = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SESSION,
                                           G_DBUS_PROXY_FLAGS_DO_NOT_CONNECT_SIGNALS |
                                           G_DBUS_PROXY_FLAGS_DO_NOT_AUTO_START,
                                           NULL, service, path, interface,
                                           NULL, &error);

    if (proxy == NULL || error)
    {
        g_warning ("g_dbus_proxy_new_for_bus_sync() failed: %s", error->message);
        return FALSE;
    }

    g_autoptr (GVariant) variant = g_dbus_proxy_get_cached_property (proxy, property);
    if (! variant)
    {
        g_warning ("Property %s.%s not found!", interface, property);
        return FALSE;
    }
    if (! g_variant_is_of_type (variant, G_VARIANT_TYPE_BOOLEAN))
    {
        g_warning ("Property %s.%s is invalid", interface, property);
        return FALSE;
    }
    
    return g_variant_get_boolean(variant);
}

int main(void)
{
    fprintf (stderr, "Status Notifier host available? %s\n",
             is_status_notifier_host_available () ? "yes" : "no");

    return 0;
}
```

Checking `IsStatusNotifierHostRegistered` property sounds to me like a good thing to do, as I'd think it not only says there's a watcher, but also an actual host.
Anyway, take this with a grain of salt, again I'm not very knowledgeable about this stuff.

Reply:
work well

Reply:
@cwendling Can I merge? I will delete libslab for the next PR and need to complete it on this basis